### PR TITLE
Objects APIs and Listeners

### PR DESCRIPTION
**Pull request by CraigLn**

### CODE REVIEW

File: .pubnub.yml [hunk 1]
```diff
@@ -103,6 +115,21 @@ features:
     - UNSUBSCRIBE-ALL
   signal:
     - SIGNAL-SEND
+  objects:
```
Comment (davidnub) [inline .pubnub.yml]: Missing `minor` version bump.

File: Sources/PubNub/Events/Subscription/ObjectEvent.swift [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+extension MembershipEvent {
+  public var description: String {
+    return ""
+  }
+}
+
+// MARK: - Implementation
+
+// extension MessageResponse: MessageEvent {
```
Comment (davidnub) [inline Sources/PubNub/Events/Subscription/ObjectEvent.swift]: Is there a reason to not just remove this commented out code?

File: Sources/PubNub/Helpers/AnyJSONType.swift [hunk 3]
```diff
@@ -373,10 +381,22 @@ extension AnyJSONType: Equatable {
...
+      if lhsArray.count == rhsArray.count {
+        for (index, lhsValue) in lhsArray.enumerated() {
+          if AnyJSONType(rawValue: lhsValue) != AnyJSONType(rawValue: rhsArray[index]) {
+            return false
+          }
+        }
+        return true
+      }
+      return false
+//      return lhs.rawArray == rhs.rawArray
```
Comment (davidnub) [inline Sources/PubNub/Helpers/AnyJSONType.swift]: Remove?

File: Sources/PubNub/Helpers/AnyJSONType.swift [hunk 4]
```diff
@@ -373,10 +381,22 @@ extension AnyJSONType: Equatable {
...
+    case let (.dictionary(lhsDict), .dictionary(rhsDict)):
+      return lhsDict.allSatisfy {
+        rhsDict[$0] == $1
+      }
+//      return lhs.rawDictionary == rhs.rawDictionary
```
Comment (davidnub) [inline Sources/PubNub/Helpers/AnyJSONType.swift]: Remove?

File: Sources/PubNub/Networking/Endpoints/SpaceObjectsEndpoint.swift [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+}
+
+struct SpaceObjectsResponseDecoder: ResponseDecoder {
+  typealias Payload = SpaceObjectsResponsePayload
+}
+
+// MARK: - Membership Response Decoder
+
+struct SpaceMembershipsObjectsResponseDecoder: ResponseDecoder {
+  typealias Payload = SpaceMembershipResponsePayload
```
Comment (davidnub) [inline Sources/PubNub/Networking/Endpoints/SpaceObjectsEndpoint.swift]: Why is the `struct` plural "Memberships" but the Response Payload singular "Membership" ?

File: Sources/PubNub/PubNubConfiguration.swift [hunk 6]
```diff
@@ -96,8 +97,17 @@ public struct PubNubConfiguration: Hashable {
...
+  public var origin: String {
+    guard let sub = subdomain else {
+      return "ps.\(domain)"
+    }
+
+    return "\(sub).\(domain)"
+  }
+
+  public var subdomain: String?
+  public var domain: String = "pubnub.com"
```
Comment (davidnub) [inline Sources/PubNub/PubNubConfiguration.swift]: Why would this not be `"pndsn.com"` by default?

File: .pubnub.yml (hunk 1 above)
Comment (CraigLn) [reply .pubnub.yml]: The version was increased from `1.1.0` to `1.2.0`

File: Sources/PubNub/Events/Subscription/ObjectEvent.swift (hunk 2 above)
Comment (CraigLn) [reply Sources/PubNub/Events/Subscription/ObjectEvent.swift]: No, just missed it.  Will delete.

File: Sources/PubNub/Helpers/AnyJSONType.swift (hunk 3 above)
Comment (CraigLn) [reply Sources/PubNub/Helpers/AnyJSONType.swift]: Yes

File: Sources/PubNub/Helpers/AnyJSONType.swift (hunk 4 above)
Comment (CraigLn) [reply Sources/PubNub/Helpers/AnyJSONType.swift]: Yes

File: Sources/PubNub/Networking/Endpoints/SpaceObjectsEndpoint.swift (hunk 5 above)
Comment (CraigLn) [reply Sources/PubNub/Networking/Endpoints/SpaceObjectsEndpoint.swift]: Yes it should just be a single.

File: Sources/PubNub/PubNubConfiguration.swift (hunk 6 above)
Comment (CraigLn) [reply Sources/PubNub/PubNubConfiguration.swift]: There was a Slack thread about how we should change to use the `subscribe key` as the sub-domain, but `pndsn.com` doesn't currently have the wildcard to allow for this.  

I can revert the change until we get more clarification on what needs to be used.

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/PubNub/swift/pullRequest?prid=4279819&bid=14600556)

File: Sources/PubNub/Events/Subscription/ObjectEvent.swift [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    let container = try decoder.container(keyedBy: CodingKeys.self)
+
+    userId = try container.decode(String.self, forKey: .userId)
+    spaceId = try container.decode(String.self, forKey: .spaceId)
+    custom = try container.decodeIfPresent([String: JSONCodableScalarType].self, forKey: .custom) ?? [:]
+    updated = try container.decodeIfPresent(Date.self, forKey: .updated) ?? Date.distantPast
+    eTag = try container.decodeIfPresent(String.self, forKey: .eTag) ?? ""
+  }
+}
+
```
Comment (madisonsmith) [inline Sources/PubNub/Events/Subscription/ObjectEvent.swift]: Codacy Issue found: [Files should have a single trailing newline.](https://app.codacy.com/app/PubNub/swift/pullRequest?prid=4279819)

File: Sources/PubNub/PubNubConfiguration.swift (hunk 6 above)
Comment (davidnub) [reply Sources/PubNub/PubNubConfiguration.swift]: Yea it was a proposal for some improvement, need to qualify the details before we go changing.

File: Sources/PubNub/Events/Subscription/ObjectEvent.swift [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+public struct ObjectSubscribePayload: Codable {
+  public let source: String
+  public let version: String
+  public let event: ObjectAction
+  public let type: ObjectType
+  public let data: AnyJSON
+}
+
+public struct IdentifierEvent: Codable, Hashable {
+  public let id: String
```
Comment (madisonsmith) [inline Sources/PubNub/Events/Subscription/ObjectEvent.swift]: Codacy Issue found: [Variable name should be between 3 and 40 characters long: 'id'](https://app.codacy.com/app/PubNub/swift/pullRequest?prid=4279819)

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/PubNub/swift/pullRequest?prid=4279819&bid=14600556)

**Pull request merged**